server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -450,14 +450,20 @@ private boolean assertConsistent() {
             // If the state is failed we have to have a reason for this failure
             assert state.failed() == false || reason != null;
             assert (state != ShardState.INIT && state != ShardState.WAITING) || nodeId != null : "Null node id for state [" + state + "]";
+            assert state != ShardState.QUEUED || (nodeId == null && generation == null && reason == null) :
+                    "Found unexpected non-null values for queued state shard nodeId[" + nodeId + "][" + generation + "][" + reason + "]";
             return true;
         }
 
-        public ShardSnapshotStatus(StreamInput in) throws IOException {
-            nodeId = in.readOptionalString();
-            state = ShardState.fromValue(in.readByte());
-            generation = in.readOptionalString();
-            reason = in.readOptionalString();
+        public static ShardSnapshotStatus readFrom(StreamInput in) throws IOException {
+            String nodeId = in.readOptionalString();
+            final ShardState state = ShardState.fromValue(in.readByte());
+            final String generation = in.readOptionalString();
+            final String reason = in.readOptionalString();
+            if (state == ShardState.QUEUED) {
```

Comment:
Not a huge win but if we have a large number of snapshots this one might be the most common status and since there's logic that compares with this status in a hot loop it's worth it IMO.